### ISSUE TITLE
Questions and Bugs

### ISSUE THREAD

**Issue by hackall360** [trickerer/Trinity-Bots#25]
I had a few questions and or bugs I felt I should address in a new issue, as I could not find them in another issue.

A few noticeable bugs that I am having with NPC Bots are:

* The faces, legs, and sometimes the tops of the shoulders, of NPC Bots without gear on, appear to have messed up textures

> While looking at the faces of the NPC Bots, they all appear to have a black cloth face, with buttons down the middle. All of them appear to have patches on their legs, like rings of textures mismatching. The NPC Bots with messed up tops of their shoulders, they just appear to have a single texture's face (like one of the faces of a cube in texturing) that is messed up in the middle. Some of the NPC Bots, specifically the Blood Elves, have Night Elf eyebrows.

#### Corrections to the above problem added, and fixed typos.

~~* If all of the NPC Bots are spawned in an area that can be loaded by players, it causes the client to crash to the login screen.~~

### Remedied in a config, **DO NOT SPAWN A BUNCH OF NPC BOTS IN ONE SPOT, OR IT WILL KICK YOU FROM THE SERVER**

> ~~When I had attempted to use a Macro script to quickly spawn all NPC Bots onto GM Island, I could no longer teleport to GM Island without my client crashing to the login screen. However, I could load in all of the NPC Bots when initially spawning all of them in, and only after leaving the area or logging out and back in, it would cause a crash. This was fixed by removing every entry 70000 - 79999 from a few different databases, and adding them back into the game in a large flattened map area, that players cannot load unless they fly into it. I was wondering if this is a client limitation, a bug in the coding, or something else.~~

#### Above problem has a workaround now.

~~* NPC Bots appear to be able to use the highest skills for their class and race, at any level.~~

### When a freestanding, or owned bot uses a spell, it is a spell they have with their level and talent tree.

~~> When creating an NPC Bot, the bots appear to be buffing players insanely, and using high ranked healing and damage spells, regardless of their level. Is this a bug, something that was intentional, or something I can change in a config?~~

#### Above problem, no-longer a problem, as I thought the NPC Bots where dealing too high of damage and such, 
 I had a misunderstanding in the differences of classes and races, until I went back through and played those classes and races again.

**Comment by trickerer** [comment]
>     * Faces or textures of most, if not all, NPC Bots appear to be garbled up.

I received such a report in the past. If you are using HD textures patch try removing it.

>     * If all of the NPC Bots are spawned in an area that can be loaded by players, it causes the client to crash to the login screen.

I never tried to spawn all of the bots in one spot so I never ran into that issue. Couple possible reasons come to mind, but I can't say for sure until I test it.
There is no such thing as 'crashing to the login screen'. If you're getting disconnected then logs could tell you something about that, or you could straight up see an error in the console.

>     * NPC Bots appear to be able to use the highest skills for their class and race, at any level.

Are you talking about free standing bots or hired ones? The only thing of all that I noticed while playing is that for some buffs lower rank is not chosen when buffing low-level target (and it's not an NPCBots issue).

**Comment by hackall360** [comment]
I have my client setup with everything default besides a patch for custom items I was creating, I can remove it and test without it, however it doesn't add anything visual, or modify any textures.

I could try to reproduce the problem with crashing, by finding where I spawned all the NPC Bots in, and go over the logs, however I didn't get any errors in the console last time, outside of `MMAP:loadMap: 5712843.mmtile was built with generator v12, expected v11` errors, due to using the latest trinitycore VMAP extraction software, instead of the one built with the version of trinitycore that NPC Bots uses, but from what I remember it would freeze and then the server could be disconnecting me due to an overload in packets, thinking that it is some kind of attack, or possibly just an error that gets thrown, and the server handles the exception by disconnecting me.

The bots that you can hire via the mercenary person, after spawning them into the world, is the bots I am talking bout. And I was meaning that the lower ranks are used when they are low levels, and their targets are low levels. If  this is not an NPC Bot issue, then is there any way to fix it?

**Comment by YunataSavior** [comment]
> NPC Bots appear to be able to use the highest ranked skills for their class and race, regardless of NPC Bot level and target level.

I've seen this occur with Priest mercenaries. If they're just standing around at level 80 and not attached to any player, then they'll cast "Shadow Protection" and "Prayer of Spirit" on players passing by, but if these players are not at a high enough level such that they can receive these buffs in the first place, then these players will receive the highest possible buff level from the bot.

> If all of the NPC Bots are spawned in an area that can be loaded by players, it causes the client to disconnect from the server, or the server to disconnect the account from game.

I've never encountered/tested this, but I do notice that when I logout/quit from my WOTLK client, it takes a lot longer to leave the game compared to normal.

**Comment by hackall360** [comment]
On the client's console, the only error for getting disconnected, is: `Gamespeed set from 0.017 to 0.017`, `NetClient::HandleDisconnect()`, `CGlueMgr::NetDisconnectHandler: Displaying script`, and the rest are proficiencies loading.

And I do not see any other errors in my server console, or server logs.

In order to reproduce the problem, start fresh on a database, or remove all spawned NPC Bots. Then manually spawn every NPC Bot in the same spot. You then have to logout and relogin to you character, and everything should disconnect from the server. I cannot figure out why this is, and there is not error logs. I can only assume it has to be a built in feature of client, to potentially avoid getting hacked or DDOSed?

**Comment by trickerer** [comment]
>>NPC Bots appear to be able to use the highest ranked skills for their class and race, regardless of NPC Bot level and target level.
>
>I've seen this occur with Priest mercenaries. If they're just standing around at level 80 and not attached to any player, then they'll cast "Shadow Protection" and "Prayer of Spirit" on players passing by, but if these players are not at a high enough level such that they can receive these buffs in the first place, then these players will receive the highest possible buff level from the bot.

Yes. I believe these are not the only spells that are affected by this bug. Target too low level check is not performed for creature caster, so if target player's level is too low to receive rank 1 of the spell, max spell rank is chosen. I encountered this in the past - just briefly - so I didn't even care to fix it at that moment.

**Comment by hackall360** [comment]
Well I will do some more digging around, playing on my private server, and see if I can encounter anymore issues, or find more information as to what the cause of my original three issues are. As far as I know, the client I have downloaded, is using no modified MPQ files, however, my client also came with all wow.exe files patched/hex edited where they can load modified MPQ files.

**Comment by trickerer** [comment]
>* If all of the NPC Bots are spawned in an area that can be loaded by players, it causes the client to disconnect from the server, or the server to disconnect the account from game.

Confirmed.
You get kicked by AntiDOS.

**Comment by hackall360** [comment]
If you get kicked by AntiDOS via the client, is there a way to force the server to serve packets out slowly, instead of sending all of the NPC Bot packets at once when attempting to load more than about 10 NPC Bots? Or is loading everybody in one area like that, something that should just be avoided entirely?

**Comment by trickerer** [comment]
You get kicked via the server and there is not way to make server think he recieved all those requests during some period of time (or that would require some serious work).
The client is sending CMSG_GET_MIRRORIMAGE_DATA packets when such units get within sight range, so problem is with '100+ bots in one spot' rather than '100+ bots in one area'.
I'll prepare a workaround which can be temporarily enabled in config for people who get in trouble like that, so they can move their bots to different areas (currently 100 packets of one type is default limit per tick).

**Comment by hackall360** [comment]
Yeah, I was noticing that went I went back to my NPC Bot spawn location, I had scattered the bots 20+ metres (In-game) away from each other for the most part, and it made it so the client cannot ask for more than like 10 or 20 bots at a time. I am assuming that every NPC Bot has its own packet that gets sent, so is there an easy way to daisy chain all packets to essentially include two or more packets per packet that is sent? Or would that require an edit to the client, to be able to handle daisy chained packets?

**Comment by trickerer** [comment]
When you move your character, your client sends movement packets to the server, the server checks what objects (including creatures) should be sent to client; if there is 101 bots in one spot server honestly sends all that info. The client sees creatures with mirror image flag set (which bots use for appearance) and sends additional requests (1 packet per bot) to the server. The server sees all those requests have the same timestamp and assumes it's a DoS attack. Many packet types has their own lmits per tick set server-side, but default is 100.
There are couple different ways this issue can be resolved. At least one of them includes client editing, but all of them seem to be very hacky to me, so a workaround is the only thing I am ready to do. Putting ALL bots in one exact spot is crazy anyways.

**Comment by hackall360** [comment]
Alright, is there any reason for my client to be rendering the NPC Bots with strange textures, instead of what I am supposed to be seeing? Because I am able to confirm that I have no patches in my client, outside of the default patches fir wrath of the lich king, and I have not changed the NPC Bots in the SQL databases at all, outside of deleting everyone from both "Spawn tables", and respawning them. I have also tried deleting the cache multiple times, and allowing the client to rebuild the cache files.

I would send a picture of what I am seeing, however my internet is not good enough to send a picture right now, it would have taken 30 minutes to upload a 5MB picture. I should have good enough internet on the 21st however, and I can try re-downloading a 3.3.5a 12340 client and see if it fixes it then, and upload an image of the problem I am having currently.

**Comment by trickerer** [comment]
The obvious thing that comes to mind is you could test other cases of creatures with mirror image flag is set - see how they appear

- Mirror Image spell itself
- That skeleton general (whatever he is called) after first boss in Halls of Reflection (copies player)
- Any other that I don't remember

**Comment by hackall360** [comment]
> * Mirror Image spell itself
I am able to test this soon, hopefully tonight.

> * That skeleton general (whatever he is called) after first boss in Halls of Reflection (copies player)
I can also try to test this out tonight, by either running the dungeon/raid or attempting to spawn the skeleton general into ScottTest map or something temporarily, and seeing if he bugs out as well.

**Comment by hackall360** [comment]
I cannot find any problems with any mirror image spells or the bosses, it appears to only be the NPC Bots that appear to have messed up textures for me.

**Comment by Parkan1986** [comment]
Can you share with your macros to spawn quick npcbots?

**Comment by trickerer** [comment]
> However, the NPC Bots would not follow me into the Icecrown Citadel Dungeon for Halls of Reflection, is this a known issue? Or is there some reason for this?

Bots must be in party and allowed to go to dungeons in config (default for non-raid dungeons)

**Comment by Parkan1986** [comment]
> When I had attempted to use a Macro script to quickly spawn all NPC Bots onto GM Island

Can you share with you macros to quikly spawn npcbots?

**Comment by Parkan1986** [comment]
Trickerer

Хотел вот скомпилировать на более свежей версии тринитикор и словил кучу ошибок.Понимаю,что Вы не так уж часто обновляете,да и думаю смысла такого частого обновления с кодом тринити нет кроме если там не введут какие-нибудь новые фичи или не сделают какие-либо фиксы.Прикладываю лог ошибок,думаю при следующем обновлении в любом случае придется исправлять.
[Новый текстовый документ.txt](https://github.com/trickerer/Trinity-Bots/files/5234076/default.txt)

**Comment by hackall360** [comment]
> > However, the NPC Bots would not follow me into the Icecrown Citadel Dungeon for Halls of Reflection, is this a known issue? Or is there some reason for this?
> 
> Bots must be in party and allowed to go to dungeons in config (default for non-raid dungeons)

My bad, forgot to copy over the settings from the last NPC Bot Update.

**Comment by hackall360** [comment]
> > When I had attempted to use a Macro script to quickly spawn all NPC Bots onto GM Island
> 
> Can you share with you macros to quikly spawn npcbots?

The macro script I used, is an XML file I generated with a java program I created. As far as I know, the macro works with my Razer Synapse software for my RGB keyboard. I cannot guarantee that it works with any other macro software.

However, the use of a macro to spawn in more than 100 NPC Bots at once, is not recommended, as per issue #25 and #1 states, it can cause a crash after reloading the NPC Bots (ie. relogging in the same spot, or coming back after unloading the map area)

[All NPCBot Spawn.zip](https://github.com/trickerer/Trinity-Bots/files/5234821/All.NPCBot.Spawn.zip)

**Comment by trickerer** [comment]
> Хотел вот скомпилировать на более свежей версии тринитикор и словил кучу ошибок.Понимаю,что Вы не так уж часто обновляете,да и думаю смысла такого частого обновления с кодом тринити нет кроме если там не введут какие-нибудь новые фичи или не сделают какие-либо фиксы.Прикладываю лог ошибок,думаю при следующем обновлении в любом случае придется исправлять.

[TC 41188eb5](https://github.com/TrinityCore/TrinityCore/commit/41188eb535e445b73909659980257da2939e4be4)

Я уже знал об этих ошибках. И в следующий раз не пиши в существующем обсуждении, пожалуйста.

**Comment by hackall360** [comment]
> I cannot find any problems with any mirror image spells or the bosses, it appears to only be the NPC Bots that appear to have messed up textures for me.

I already tested this once, summoning the boss, however I am going to test this again by running through the dungeon with NPC Bots, to see if the instanced boss reacts differently.

**Comment by Parkan1986** [comment]
> > Хотел вот скомпилировать на более свежей версии тринитикор и словил кучу ошибок.Понимаю,что Вы не так уж часто обновляете,да и думаю смысла такого частого обновления с кодом тринити нет кроме если там не введут какие-нибудь новые фичи или не сделают какие-либо фиксы.Прикладываю лог ошибок,думаю при следующем обновлении в любом случае придется исправлять.
> 
> [TC 41188eb5](https://github.com/TrinityCore/TrinityCore/commit/41188eb535e445b73909659980257da2939e4be4)
> 
> Я уже знал об этих ошибках. И в следующий раз не пиши в существующем обсуждении, пожалуйста.

Извиняюсь. В следующий раз буду создавать отдельный топик

**Comment by 328950225** [comment]
1.I think bot_warlock is not an auxiliary class, so when BOT_SPEC_WARLOCK_DEMONOLOGY, the first SummonBotPet is BOT_PET_FELGUARD, the bot is lv60;
2,When nocombat maybe did't use heal hot is rihgt;
3,After i converting to Chinese ,the weapons Enchant "msg1" is empty,ingame gossipmenu is  empty .i don't know what's wrong,like this :
`        bool HasAbilitiesSpecifics() const override { return true; }
        void FillAbilitiesSpecifics(std::list<std::string> &specList) override
        {
            std::ostringstream msg1;
            msg1 << "主手: ";
            switch (mhEnchant)
            {
            case CRIPPLING_POISON_1:    msg1 << "减速药膏";break;
            case INSTANT_POISON_1:      msg1 << "速效药膏";break;
            case DEADLY_POISON_1:       msg1 << "致命药膏";break;
            case WOUND_POISON_1:        msg1 << "致伤药膏";break;
            case MIND_NUMBING_POISON_1: msg1 << "麻痹药膏";break;
            case ANESTHETIC_POISON_1:   msg1 << "麻醉药膏";break;
            default:                    msg1 << "没有涂抹毒药";break;
            }
            specList.push_back(msg1.str());

            std::ostringstream msg2;
            msg2 << "副手: ";
            switch (ohEnchant)
            {
            case CRIPPLING_POISON_1:    msg2 << "减速药膏";break;
            case INSTANT_POISON_1:      msg2 << "速效药膏";break;
            case DEADLY_POISON_1:       msg2 << "致命药膏";break;
            case WOUND_POISON_1:        msg2 << "致伤药膏";break;
            case MIND_NUMBING_POISON_1: msg2 << "麻痹药膏";break;
            case ANESTHETIC_POISON_1:   msg2 << "麻醉药膏";break;
            default:                    msg2 << "没有涂抹毒药";break;
            }
            specList.push_back(msg2.str());
        }`

**Comment by trickerer** [comment]
1. So, you think Felguard must be a number one priority for Demonology Warlocks? I personally have no idea, I never played that class for real, but it can be easily done. In any case, here is some explanation why it is not currently: non-controlled warlock bot will choose random pet regardless of spec, and will keep it forever unless said otherwise; pet type is not saved between server restarts.
2. Are you saying that using HoTs to heal out of combat is not the best idea? I don't see why. Anyways, again, here is a bit of explanation: healing spell choice is based purely on (expected) health loss, so if target is almost at full health choice between Flash Heal and Renew for example, is semi-random. Mana effeciency and cast time are not taken into consideration here.
3. What about other strings? I never tried to translate this text in source, let alone to translate it into Chinese. There is no localization support as of yet, but I'm interested.

**Comment by 328950225** [comment]
> 1. So, you think Felguard must be a number one priority for Demonology Warlocks? I personally have no idea, I never played that class for real, but it can be easily done. In any case, here is some explanation why it is not currently: non-controlled warlock bot will choose random pet regardless of spec, and will keep it forever unless said otherwise; pet type is not saved between server restarts.
> 2. Are you saying that using HoTs to heal out of combat is not the best idea? I don't see why. Anyways, again, here is a bit of explanation: healing spell choice is based purely on (expected) health loss, so if target is almost at full health choice between Flash Heal and Renew for example, is semi-random. Mana effeciency and cast time are not taken into consideration here.
> 3. What about other strings? I never tried to translate this text in source, let alone to translate it into Chinese. There is no localization support as of yet, but I'm interested.

in full health an not incombat ,but the bot always giveing the heal hot to player or other bot .

**Comment by Parkan1986** [comment]
> > 1. So, you think Felguard must be a number one priority for Demonology Warlocks? I personally have no idea, I never played that class for real, but it can be easily done. In any case, here is some explanation why it is not currently: non-controlled warlock bot will choose random pet regardless of spec, and will keep it forever unless said otherwise; pet type is not saved between server restarts.
> > 2. Are you saying that using HoTs to heal out of combat is not the best idea? I don't see why. Anyways, again, here is a bit of explanation: healing spell choice is based purely on (expected) health loss, so if target is almost at full health choice between Flash Heal and Renew for example, is semi-random. Mana effeciency and cast time are not taken into consideration here.
> > 3. What about other strings? I never tried to translate this text in source, let alone to translate it into Chinese. There is no localization support as of yet, but I'm interested.
> 
> if it's auto ,and spec is Demonology ,so auto summon BOT_PET_FELGUARD,this is first action.
> in full health and not incombat ,but the bot always giveing the heal hot to player or other bot .
> other strings is OK.

Sorry for interrupting your talk. I have question to 328950225:
1.I am looked to your for of npcbots and got few question. I saw that you add something like random item stats. Whatthis patch do? Is this the same like in wow with items random suffixes or random property? Can you describe it? 
2.if i decide to compile from your source, what i need to remove or change to not get chinese text in game? Does sql files from trinity will work if i change to them not from your source? 
3.any possibility to talk to you in live? Discord or some other programs to chat?

**Comment by trickerer** [comment]
@328950225 next time create a new issue instead.

**Comment by 328950225** [comment]
@Parkan1986 when you buy equipment from vendor .this  random item stats can enchanting equipment.
About localization，when you patched npcbot4.8.4a.You can just download all the files in Src / game /AI/ npcbot Replace it

**Comment by trickerer** [comment]
This issue is mostly solved, and now is a mess. If remaining problem with textures persists, please create a separate issue for that.
